### ISSUE TITLE
[Admin Support] Delete account will surge up scheduled jobs incredibly

### ISSUE THREAD

**Issue by aka-rin** [tootsuite/mastodon#10476]
### What happened?

Yesterday I deleted hundreds of fake accounts on my instance. Most of them were registered in August, 2018. But today the amount of scheduled jobs in Sidekiq make a drastic increase -- may be the highest record in the history.

So I made a quick investigation and it turns out that most of them are [delete activity](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-delete) that send to inboxes of different servers with a tag named `ActivityPub::LowPriorityDeliveryWorker`. And then I found it in the source code:

> https://github.com/tootsuite/mastodon/blob/master/app/services/suspend_account_service.rb#L115

If I didn't get it wrong, it means that if you delete an account on your instance, an individual delete activity will be sent to all the instances that (once) connect to your instance. If you delete 100 accounts and your instance have connections with 10000 instances, the total requests amount will be 100*10000 = 1 million ...

That's absurd. Most of these accounts are totally empty - no toots, no followers and no interactions. But if you want to delete them, you have to notice every instance you know and waste lots of memory and traffic. I can't imagine what happen if @Gargron delete 100 accounts on mastodon.social 😭

**Comment by ealgase** [comment]
What would you suggest as an alternative?  Storing every instance that has searched for a given user would be a pain (and somewhat against Mastodon's mission of "no user tracking"), so unfortunately, this seems to be the best solution.

**Comment by aka-rin** [comment]
There are many ways to prevent this situation, for example:

- Send multiple activities at once.
- Keep a record of cross-instance interactions for each account.
- Make a delete state for deleted accounts in cross-instance interactions.
- ...

However, most of them need to modify the protocol and I am afraid that it will take a lifetime to wait ...

**Comment by jeroenpraat** [comment]
Related: https://framagit.org/DavidLibeau/activitypubactor/issues/11

**Comment by DavidLibeau** [comment]
It seems that Mastodon instances send me (at http://twitter.activitypub.actor/) a very high number of request (40k/min) which kind of DDOS my server. I don't know if this issue is related.

**Comment by aka-rin** [comment]
@DavidLibeau Yes, it is definitely the same problem. I just feel a bit disappointed ...

**Comment by salvadorpla** [comment]
Yes, this is annoying. Every time a local spam user is deleted my server's log is filled up quickly with thousands of Job Raised exception warnings:
 
WARN: {"context":"Job raised exception","job":{"class":"ActivityPub::LowPriorityDeliveryWorker", "args":["{\"@context\":\"https://www.w3.org/ns/activitystreams\",\"id\":\"https://mastodont.cat/users/fpttelecomhcmvn#delete\",\"type\":\"Delete\",...

followed by:

WARN: HTTP::TimeoutError: Delivery failed for https://twitter.activitypub.actor/fpttelecomhcmvn/inbox: execution expired on https://twitter.activitypub.actor/fpttelecomhcmvn/inbox

Plenty of above lines every time you delete an user:
cat /var/log/messages | grep 'Delivery failed for https://twitter.activitypub.actor' |grep 'Jul  2' | wc -l
23750

23750 log lines of crap after the deletion of a single user!?
We need to stop this urgently. I don't want my Mastodon server talking to any twitter alike service, at all.
@Gargron @DavidLibeau

**Comment by Gargron** [comment]
The http://twitter.activitypub.actor/ inbox returns a 501 error but the service advertises having an inbox and there is no mechanism that says that a 501 error should mean the advertised inbox is permanently unavailable. It doesn't help that you don't have a shared inbox but individual URLs for each user so they all receive a message each.

**Comment by Gargron** [comment]
It would a) reduce amount of requests for the same message b) help the failure tracker come to the conclusion the URL is permanently inaccessible sooner. Ideally you wouldn't have to have an inbox URL at all but I'm not sure how well-supported that case is in the code--at some point it was a requirement to have an inbox, and that might still be so.

**Comment by DavidLibeau** [comment]
Ok, that's what I though. I temporary removed all inbox endpoint references.

**Comment by strypey** [comment]
Nob question: would it be possible to have the instance that has deleted the account send delete notifications only when another instance tries to interact with it?

**Comment by kescherCode** [comment]
I think some things to remedy this long-standing issue could be done:

- Make this a really, *really* low priority job
- Batch up account deletions
- Intentionally make account deletion federate very slowly.
